File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/dao/PersonDao.java
package com.dani_chuks.andeladeveloper.starwars.data.db.local.dao;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Person;

import java.util.List;

import io.reactivex.Flowable;

@Dao
public interface PersonDao {
    @Query("SELECT * FROM person")
    Flowable<List<Person>> getAll();

    @Query("SELECT * FROM person LIMIT :limit")
    Flowable<List<Person>> getItemBySize(int limit);

    @Query("SELECT * FROM person ORDER BY name ASC" )
    Flowable<List<Person>> getAllAlphabetically();

    @Query("SELECT * FROM person WHERE gender = :gender")
    Flowable <List<Person>> getPeopleByGender(String gender);

    @Query("SELECT * FROM person WHERE url = :personUrl")
    Flowable<Person> getPersonByURL(String personUrl);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updatePeople(List<Person> people);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertPeople(List<Person> people);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertPerson(Person person);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updatePerson(Person person);
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/dao/PersonDao.kt
package com.dani_chuks.andeladeveloper.starwars.data.db.local.dao

import androidx.room.Dao
import androidx.room.Query
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Person
import kotlinx.coroutines.flow.Flow

@Dao
abstract class PersonDao: BaseDao<Person> {
    @Query("SELECT * FROM person")
    abstract suspend fun all(): List<Person>

    @Query("SELECT * FROM person ORDER BY name ASC")
    abstract suspend fun allAlphabetically(): List<Person>

    @Query("SELECT * FROM person LIMIT :limit")
    abstract suspend fun getItemBySize(limit: Int): List<Person>

    @Query("SELECT * FROM person WHERE gender = :gender")
    abstract suspend fun getPeopleByGender(gender: String): List<Person>

    @Query("SELECT * FROM person WHERE url = :personUrl")
    abstract suspend fun getPersonByURL(personUrl: String): Person?

    @Query("SELECT * FROM person")
    abstract fun allAsFlow(): Flow<List<Person>?>

    @Query("SELECT * FROM person ORDER BY name ASC")
    abstract fun allAlphabeticallyAsFlow(): Flow<List<Person>?>

    @Query("SELECT * FROM person LIMIT :limit")
    abstract fun getItemBySizeAsFlow(limit: Int): Flow<List<Person>?>

    @Query("SELECT * FROM person WHERE gender = :gender")
    abstract fun getPeopleByGenderAsFlow(gender: String): Flow<List<Person>?>

    @Query("SELECT * FROM person WHERE url = :personUrl")
    abstract fun getPersonByURLAsFlow(personUrl: String): Flow<Person?>
}
